Classes/Middleware/OAuth2Authorization.php

```diff
@@ -60,40 +58,50 @@ public function process(ServerRequestInterface $request, RequestHandlerInterface
         $factory = new ServerFactory();
         $server = $factory->buildAuthorizationServer();
 
-        $userSession = new UserSession();
+        $frontendUser = $request->getAttribute('frontend.user');
+
+        $userSession = new UserSession($frontendUser);
         $authorizationRequest = $userSession->getData('oauth2.authorizationRequest');
 
-        $this->bootFrontendController();
+        $router = $this->siteFinder->getSiteByPageId($this->configuration->getLoginPage())->getRouter($this->context);
 
         if (!$authorizationRequest) {
             try {
                 $authorizationRequest = $server->validateAuthorizationRequest($request);
             } catch (OAuthServerException $e) {
-
-                $redirectUri = $this->getContentObjectRenderer()->typoLink_URL([
-                    'parameter' => sprintf('t3://page?uid=%d', $this->configuration->getLoginPage()),
-                ]);
-
                 $this->logger->warning(sprintf('Validating authorization request failed: %s', $e->getMessage()));
 
+                $redirectUri = $router->generateUri($this->configuration->getLoginPage());
+
                 return new RedirectResponse($redirectUri);
             }
         }
 
-        /** @var Context */
-        $context = GeneralUtility::makeInstance(Context::class);
+        if (!$this->context->getPropertyFromAspect('frontend.user', 'isLoggedIn', false)) {
+            $userSession->setData('oauth2.authorizationRequest', serialize($authorizationRequest));
+
+            $redirectUri = $router->generateUri($this->configuration->getLoginPage(), ['redirect_url' => $request->getUri()->getPath()]);
 
-        if (!$context->getPropertyFromAspect('frontend.user', 'isLoggedIn', false)) {
-            $userSession->setData('oauth2.authorizationRequest', $authorizationRequest);
+            return new RedirectResponse($redirectUri);
+        }
 
-            $redirectUri = $this->getContentObjectRenderer()->typoLink_URL([
-                'parameter' => sprintf('t3://page?uid=%d&redirect_url=%s', $this->configuration->getLoginPage(), $request->getUri()->getPath()),
-            ]);
+        // With TYPO3 11.5.17 it takes 3 loops to unserialize the AuthorizationRequest
```

Comment:
This sounds a bit weird. I would love to have a more detailed comment here how the data looks like and why it is encoded multiple times (?)

Reply:
I'm not 100% sure where it does come from. But the session Data serialization in the core was changed from 10.4 to 11.5.  An the core serializes the data also. So it seems to be serialized multiple times.